{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import time\n",
    "import gzip\n",
    "from concurrent import futures\n",
    "from functools import partial\n",
    "import pandas as pd\n",
    "from CHARMio import parse_pairs, parse_gtf, write_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cli(args)->int:\n",
    "    filename, gtf_file, out_name, thread = \\\n",
    "        args.filename[0], args.gtf_filename, args.out_name, args.num_thread\n",
    "    # parsing ref gtf and pairs file\n",
    "    pairs = parse_pairs(filename)\n",
    "    # build in-mem exon index\n",
    "    gtf = parse_gtf(gtf_file)\n",
    "    ref = build_in_memory_index(get_exon(gtf))\n",
    "    # do search\n",
    "    cleaned = clean_splicing(pairs, ref, thread)\n",
    "    write_pairs(cleaned, out_name)\n",
    "def get_exon(gtf:pd.DataFrame) -> pd.DataFrame:\n",
    "    # extract exon-gene_name from gtf table\n",
    "    relevant = gtf.query('feature == \"exon\" & source == \"HAVANA\"') #using HAVANA only\n",
    "    gene_id = relevant[\"group\"].str.extract('gene_id \"(ENSG[0-9]{11}.[0-9])\";') #extract gene name from group\n",
    "    gene_id.columns = [\"gene_id\"] # extract returns dataframe rather than series\n",
    "    # don't mind strand\n",
    "    return pd.concat([relevant.drop([\"group\",\"feature\",\"source\",\"score\",\"strand\",\"frame\"],axis=1),gene_id],axis=1)\n",
    " \n",
    "def build_in_memory_index(exons:pd.DataFrame) -> dict:\n",
    "    # split by chr and using IntervalIndex to enable searching\n",
    "    ref_dict = {key : value for key, value in exons.groupby(\"seqname\")}\n",
    "    # build index by chromosome\n",
    "    for chromosome in ref_dict:\n",
    "        # using start, end attrs as index\n",
    "        by_chr_table = ref_dict[chromosome]\n",
    "        bed_tuple = by_chr_table.set_index(['start','end']).index \n",
    "        bed_interval_index = pd.IntervalIndex.from_tuples(bed_tuple)\n",
    "        by_chr_table.index = bed_interval_index\n",
    "        ref_dict[chromosome] = by_chr_table.drop([\"start\",\"end\",\"seqname\"],axis=1)\n",
    "    sys.stderr.write(\"hires_utils::clean_splicing: index done.\\n\")\n",
    "    return ref_dict\n",
    "def legs_co_gene(contact:pd.Series, chromosome:str, ref_dict:dict)->bool:\n",
    "    # whether two legs of contacts are in same gene's exon\n",
    "    # must be intra contacts\n",
    "    pos1, pos2 = contact[\"pos1\"], contact[\"pos2\"]\n",
    "    result = ref_dict[chromosome][ref_dict[chromosome].index.contains(pos1)]\n",
    "    if len(result[result.index.contains(pos2)]) > 0:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "def search_chromosome(contacts_at_chromosome:tuple, ref:dict)->pd.DataFrame:\n",
    "    # search whole chromosome using F::legs_co_gene\n",
    "    # single chr searching for multi-process calling\n",
    "    chromosome, contacts = contacts_at_chromosome[0], contacts_at_chromosome[1]\n",
    "    hit_index = contacts.apply(legs_co_gene, chromosome=chromosome, ref_dict=ref, axis=1)\n",
    "    return contacts[hit_index]\n",
    "\n",
    "def clean_splicing(pairs:pd.DataFrame, ref:dict, thread:int)->pd.DataFrame:\n",
    "    '''\n",
    "    clean contacts from splicing\n",
    "    '''\n",
    "    #intra = pairs.query('chr1 == chr2') # only search for intra\n",
    "    intra = pairs.loc[pairs['chr1'] == pairs['chr2']]\n",
    "    working_func = partial(search_chromosome, ref=ref) # pool.map can't take multiple iterable as arguments\n",
    "    input_data = [(chromosome, per_chr_contacts) for chromosome, per_chr_contacts in intra.groupby(\"chr1\")] # pool.map can't take additional keyword argument\n",
    "    sys.stderr.write(\"hires_utils::clean_splicing: input parsed, search in %d thread\\n\" % thread)\n",
    "    # do multi-thread searching\n",
    "    with futures.ProcessPoolExecutor(thread) as pool:\n",
    "        res = pool.map(working_func, input_data)\n",
    "    result = pd.concat(res, axis=0) # contained contacts\n",
    "    cleaned = pairs.drop(result.index, axis=0) # clean contacts\n",
    "    print(\"clean_splicing: %d contacts removed in %s\\n\" %(len(result), pairs.attrs[\"name\"]) )\n",
    "    return cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_file = \"/share/Data/public/ref_genome/rabbit_ref/rawdata/genomic.gtf\"\n",
    "gtf_1 = \"/share/Data/public/ref_genome/human/rawdata/genomic.gtf\"\n",
    "gtf = parse_gtf(gtf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rabbit'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.search(r'rabbit', gtf_file).group(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_exon_rab(gtf:pd.DataFrame) -> pd.DataFrame:\n",
    "    # extract exon-gene_name from gtf table\n",
    "    relevant = gtf.query('feature == \"exon\"') #using HAVANA only\n",
    "    gene_id = relevant[\"group\"].str.extract('gene_id \"([A-Za-z0-9_-]+)\";') #extract gene name from group\n",
    "    gene_id.columns = [\"gene_id\"] # extract returns dataframe rather than series\n",
    "    # don't mind strand\n",
    "    return pd.concat([relevant.drop([\"group\",\"feature\",\"source\",\"score\",\"strand\",\"frame\"],axis=1),gene_id],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>4619</td>\n",
       "      <td>31619</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SH2D3C\"; transcript_id \"\"; db_xref \"G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4619</td>\n",
       "      <td>31619</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SH2D3C\"; transcript_id \"XM_051835448....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>4619</td>\n",
       "      <td>4777</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SH2D3C\"; transcript_id \"XM_051835448....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>6191</td>\n",
       "      <td>6665</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SH2D3C\"; transcript_id \"XM_051835448....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>15918</td>\n",
       "      <td>15957</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SH2D3C\"; transcript_id \"XM_051835448....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051336</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15314</td>\n",
       "      <td>15379</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"KEF51_t21\"; transcript_id \"unassigned...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051337</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15314</td>\n",
       "      <td>15379</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"KEF51_t21\"; transcript_id \"unassigned...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051338</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>15380</td>\n",
       "      <td>15445</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"KEF51_t22\"; transcript_id \"\"; db_xref...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051339</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15380</td>\n",
       "      <td>15445</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"KEF51_t22\"; transcript_id \"unassigned...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051340</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15380</td>\n",
       "      <td>15445</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"KEF51_t22\"; transcript_id \"unassigned...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2051341 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqname  source     feature  start    end score strand frame  \\\n",
       "0        NC_067374.1  Gnomon        gene   4619  31619     .      +     .   \n",
       "1        NC_067374.1  Gnomon  transcript   4619  31619     .      +     .   \n",
       "2        NC_067374.1  Gnomon        exon   4619   4777     .      +     .   \n",
       "3        NC_067374.1  Gnomon        exon   6191   6665     .      +     .   \n",
       "4        NC_067374.1  Gnomon        exon  15918  15957     .      +     .   \n",
       "...              ...     ...         ...    ...    ...   ...    ...   ...   \n",
       "2051336  NC_001913.1  RefSeq  transcript  15314  15379     .      +     .   \n",
       "2051337  NC_001913.1  RefSeq        exon  15314  15379     .      +     .   \n",
       "2051338  NC_001913.1  RefSeq        gene  15380  15445     .      -     .   \n",
       "2051339  NC_001913.1  RefSeq  transcript  15380  15445     .      -     .   \n",
       "2051340  NC_001913.1  RefSeq        exon  15380  15445     .      -     .   \n",
       "\n",
       "                                                     group  \n",
       "0        gene_id \"SH2D3C\"; transcript_id \"\"; db_xref \"G...  \n",
       "1        gene_id \"SH2D3C\"; transcript_id \"XM_051835448....  \n",
       "2        gene_id \"SH2D3C\"; transcript_id \"XM_051835448....  \n",
       "3        gene_id \"SH2D3C\"; transcript_id \"XM_051835448....  \n",
       "4        gene_id \"SH2D3C\"; transcript_id \"XM_051835448....  \n",
       "...                                                    ...  \n",
       "2051336  gene_id \"KEF51_t21\"; transcript_id \"unassigned...  \n",
       "2051337  gene_id \"KEF51_t21\"; transcript_id \"unassigned...  \n",
       "2051338  gene_id \"KEF51_t22\"; transcript_id \"\"; db_xref...  \n",
       "2051339  gene_id \"KEF51_t22\"; transcript_id \"unassigned...  \n",
       "2051340  gene_id \"KEF51_t22\"; transcript_id \"unassigned...  \n",
       "\n",
       "[2051341 rows x 9 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>4619</td>\n",
       "      <td>4777</td>\n",
       "      <td>SH2D3C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>6191</td>\n",
       "      <td>6665</td>\n",
       "      <td>SH2D3C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>15918</td>\n",
       "      <td>15957</td>\n",
       "      <td>SH2D3C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>21189</td>\n",
       "      <td>21317</td>\n",
       "      <td>SH2D3C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_067374.1</td>\n",
       "      <td>22568</td>\n",
       "      <td>23019</td>\n",
       "      <td>SH2D3C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051323</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>13579</td>\n",
       "      <td>14103</td>\n",
       "      <td>KEF51_p02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051329</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>14104</td>\n",
       "      <td>14171</td>\n",
       "      <td>KEF51_t20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051332</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>14175</td>\n",
       "      <td>15313</td>\n",
       "      <td>KEF51_p01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051337</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>15314</td>\n",
       "      <td>15379</td>\n",
       "      <td>KEF51_t21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2051340</th>\n",
       "      <td>NC_001913.1</td>\n",
       "      <td>15380</td>\n",
       "      <td>15445</td>\n",
       "      <td>KEF51_t22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>979053 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqname  start    end    gene_id\n",
       "2        NC_067374.1   4619   4777     SH2D3C\n",
       "3        NC_067374.1   6191   6665     SH2D3C\n",
       "4        NC_067374.1  15918  15957     SH2D3C\n",
       "5        NC_067374.1  21189  21317     SH2D3C\n",
       "6        NC_067374.1  22568  23019     SH2D3C\n",
       "...              ...    ...    ...        ...\n",
       "2051323  NC_001913.1  13579  14103  KEF51_p02\n",
       "2051329  NC_001913.1  14104  14171  KEF51_t20\n",
       "2051332  NC_001913.1  14175  15313  KEF51_p01\n",
       "2051337  NC_001913.1  15314  15379  KEF51_t21\n",
       "2051340  NC_001913.1  15380  15445  KEF51_t22\n",
       "\n",
       "[979053 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_exon_rab(gtf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hic2",
   "language": "python",
   "name": "hic2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
